### ISSUE TITLE
Cannot provision Azure Cloud Services using Terraform and ARM Template

### ISSUE THREAD

**Issue by ktenediosmyob** [terraform-providers/terraform-provider-azurerm#2541]
Hi there, up until 12 December 2018, my team was able to provision Cloud Services using Terraform and ARM Templates. However, my team has started to notice on 14 December that after Terraform generates a plan listing out the infrastructure changes required, and my team approves the plan, the Terraform job fails with an error message similar to the following:

```
azurerm_template_deployment.cloud_service: Error creating deployment: resources.DeploymentsClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="InvalidTemplate" Message="Deployment template validation failed: 'The template resource '[concat('deploy-', variables('names')[copyIndex()])]' at line '1' and column '340' is invalid. The api-version '2016-02-01' used to deploy the template does not support 'ResourceGroup' property. Please use api-version '2017-05-10' or later to deploy the template. Please see https://aka.ms/arm-template/#resources for usage details.'."
```

This is happening for all of the services that my team provisions Cloud Services in all Azure subscriptions (under all Azure Active Directory tenants) that my team supports.

My team is seeing this error when executing the job through our Jenkins instance, as well as when debugging this issue on our laptops.

For our Jenkins instance, the version of Terraform and the AzureRM Provider has not changed in months, and neither have our Terraform configs for the services that we deploy to Azure.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

As used by Jenkins:
Terraform v0.11.8
AzureRM Provider v1.12

As used on my machine:
Terraform v0.11.11
AzureRM Provider v1.20.0

### Affected Resource(s)

* azurerm_template_deployment

### Terraform Configuration Files

Complete set of Terraform config files for provisioning Cloud Services:
[cloudservice.zip](https://github.com/terraform-providers/terraform-provider-azurerm/files/2692378/cloudservice.zip)

Should the above link not work, below are a couple of files of interest.

TPL file containing ARM template for provisioning Cloud Services:

```
{
  "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
    "location": {
      "type": "string",
      "metadata": {
        "description": "The region to create cloud service"
      }
    }
  },
  "variables": {
    "names": ${names},
    "ids": ${id_lookups}
  },
  "resources": [
    {
      "apiVersion": "2018-02-01",
      "name": "[concat('deploy-', variables('names')[copyIndex()])]",
      "type": "Microsoft.Resources/deployments",
      "resourceGroup": "[variables('names')[copyIndex()]]",
      "properties": {
        "mode": "Incremental",
        "template": {
          "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
          "contentVersion": "1.0.0.0",
          "parameters": {},
          "variables": {},
          "resources": [
            {
              "apiVersion": "2017-11-15",
              "type": "Microsoft.ClassicCompute/domainNames",
              "name": "[variables('names')[copyIndex()]]",
              "location": "[parameters('location')]"
            }
          ]
        }
      },
      "copy": {
        "name": "cloud_service_copy",
        "count": "[length(variables('names'))]"
      }
    }
  ],
  "outputs": {
    "ids": {
      "type": "string",
      "value": "[string(variables('ids'))]"
    }
  }
}

```

provider.tf:

```
# introduction of Local Values configuration language feature
terraform {
  required_version = ">= 0.10.3"
}

provider "template" {
  version = "~> 1.0"
}
```

### Debug Output

https://gist.github.com/ktenediosmyob/a7e71f506bf1735817fd86c09b7ae456

The above gist consists of three files:

* `plan.log` - the output generated by Terraform when generating the plan, before the plan was applied.
* `apply-plan.log` - the output generated by Terraform when applying the plan.
* `last-successful-run.log` - the Jenkins log output of the last time Cloud Services were successfully provisioned using Terraform and ARM templates.

### Expected Behavior

That the generated plan was applied successfully, as used to be the case up until 12 December 2018.

### Actual Behavior

The following error is generated when applying the plan:

```
azurerm_template_deployment.cloud_service: Error creating deployment: resources.DeploymentsClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="InvalidTemplate" Message="Deployment template validation failed: 'The template resource '[concat('deploy-', variables('names')[copyIndex()])]' at line '1' and column '340' is invalid. The api-version '2016-02-01' used to deploy the template does not support 'ResourceGroup' property. Please use api-version '2017-05-10' or later to deploy the template. Please see https://aka.ms/arm-template/#resources for usage details.'."
```

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

* Deploying Cloud Services to Australia East and Australia Southeast.
* Issue occurs for all of the Cloud Services in all of the Azure subscriptions and Azure Active Directory tenants supported by my team.

### References

Although the following link appears to be a separate issue, perhaps these issues are related?
https://github.com/terraform-providers/terraform-provider-azurerm/issues/2507

**Comment by tombuildsstuff** [comment]
hey @ktenediosmyob 

Thanks for opening this issue :)

Taking a quick look into this this appears to be a duplicate of #2507 - looking at the Provider block defined above:

```
terraform {
  required_version = ">= 0.10.3"
}

provider "template" {
  version = "~> 1.0"
}
```

Whilst the versions of Terraform Core and the Template provider are being pinned - the version of the AzureRM Provider isn't, which is why version 1.20 is being used:

> 2018/12/18 15:24:36 [DEBUG] found valid plugin: "azurerm", "1.20.0", "/Users/kosta.tenedios/Documents/workspace/git/AccountRightDeploy/terraform/ar-api/.terraform/plugins/darwin_amd64/terraform-provider-azurerm_v1.20.0_x4"

Whilst the underlying cause of #2507 has been fixed, it's not been released yet - as such you'll need to pin to an older version of the Provider for the moment, unfortunately. You can do this via:

```
provider "azurerm" {
  version = "=1.19.0"
}
```

Also, I'm sure you're aware but just to note [that Azure Cloud Services are deprecated in favour of Azure App Services](https://github.com/MicrosoftDocs/azure-docs/issues/11861#issuecomment-427452081) (which are [natively supported by Terraform](https://www.terraform.io/docs/providers/azurerm/r/app_service.html)) - so I'm wondering if using that would potentially be an option instead? :)

Since the underlying issue for this has been fixed in #2507 - I'm going to close this issue for the moment (however we'll continue responding 😄)

Thanks!

**Comment by ktenediosmyob** [comment]
Hi Tom,

Thanks for the response.

Yesterday my team lead (Alan Prescilla) and I also came to the conclusion that we need to pin an older version of the AzureRM provider. When I tested this locally on my machine, it worked. I am now testing this in our Jenkins instance, using Terraform 0.11.11.

Yes, my team is aware of the talk around deprecating Cloud Services. At this point in time, moving to App Services would be a significant amount of work that would be done by at least one other team in my organisation. The reason why we are using Cloud Services is due to our architecture and supporting certain workflows that require tooling that cannot be deployed to App Services.

Regards,

Kosta Tenedios
DevOps Engineer

167 Cremorne St, Richmond
Melbourne VIC 3121
www.myob.com<https://www.myob.com>
[MYOB Simplify Success]<https://www.myob.com/>
